### PR TITLE
modify existing code to use newly introduced classes.

### PR DESCRIPTION
**Pull request by mfrisbey**
## Description

This PR builds on the new classes that were introduced in the previous [PR](https://github.com/adobe/aem-upload/pull/29). Here we're modifying existing classes to instead use functionality that is now provided elsewhere.

The strategy for maximizing the performance of recursively uploading directories will be as follows:

* Walk the entire directory structure of the path to be uploaded.
* As each directory is walked, build a list of all files in the directory.
* Build a list of all parts for all files in the directory.
* Add all the parts to a "concurrent queue" that will continually and concurrently upload parts, up to a maximum number being uploaded at once.

This change moves closer to implementing that strategy by:

* Modifying the existing `DirectBinaryUploadProcess` class so that it can be reused on multiple directories throughout the overall upload process, instead of _driving_ the upload process.
* `DirectBinaryUploadProcess` now makes use of the `PartUploader`, which serves as a centralized "queue" that all `DirectBinaryUploadProcess` instances will use for uploading file parts.
* `DirectBinaryUploadProcess` now takes advantage of `FileTransferHandler` to communicate changes in file upload status.
* `DirectBinaryUploadProcess` now uses the new `HttpClient` class to submit HTTP requests.

## Related Issue

Please see #20 for corresponding feature request.

## Motivation and Context

There have been several requests for this feature. It simplifies the process of uploading complex directory structures to AEM by allowing users to upload a local directory, which will be mirrored in AEM.

## How Has This Been Tested?

This applies to all the PRs that I'll submit for this feature. There is an existing set of unit tests, all of which are passing. I also added new tests to verify the new functionality that is introduced. In addition, I've created end-to-end tests that I manually ran against a local AEM instance to ensure the process works as expected.

## Screenshots (if appropriate):

N/A

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/direct-binary-upload-process.js [hunk 1]
```diff
@@ -12,238 +12,159 @@ governing permissions and limitations under the License.
...
+    async upload(uploadResult) {
+        const initResponse = await this.initiateUpload(uploadResult);
+        this.completeUri = initResponse.getCompleteUri();
+
+        await this.partUploader.uploadParts(uploadResult, initResponse.getAllParts(), this);
+
+        uploadResult.stopTimer();
+
+        // output json result to logger
+        this.logInfo('Uploading result in JSON: ' + JSON.stringify(uploadResult, null, 4));
```
Comment (tmathern) [inline src/direct-binary-upload-process.js:96]: If there are secrets, or presigned urls in the `uploadResult` object, you'll need to redact them.

File: src/direct-binary-upload-process.js [hunk 2]
```diff
@@ -12,238 +12,159 @@ governing permissions and limitations under the License.
...
-            this.initCancelToken = createCancelToken();
-            const response = await timedRequest(initOptions, this.getRetryOptions(options, uploadResult), this.initCancelToken);
-            this.initCancelToken = null;
-            const {
-                data: resObj,
-                status: statusCode,
-                elapsedTime = 0,
-            } = response;
-            options = updateOptionsWithResponse(options, response);
+        this.logInfo(`Initiating upload to ${url}`);
```
Comment (tmathern) [inline src/direct-binary-upload-process.js:130]: I would make this a debug log - logging presigned url is frowned upon in prod (we got it as feedback in Nui security review - we shouldn't log presigned urls).

**Review by tmathern — APPROVED** [review]

File: test/upload-result.test.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+const { importFile } = require('./testutils');
+
+const UploadResult = importFile('upload-result');
+const { setTimeoutPromise } = importFile('utils');
+
+describe('UploadResult Tests', function() {
+    it('test timer', async function() {
+        const uploadResult = new UploadResult();
+        uploadResult.startTimer();
+        await setTimeoutPromise(20);
```
Comment (tmathern) [inline test/upload-result.test.js:24]: FYI, you can control the clock in tests using `sinon clock`, so you don't have to use sleeps/timeouts.
See documentation for details: https://sinonjs.org/releases/v1.17.6/fake-timers/

There are some nui tests that use the sinon clock to control time and verify timers. I can dig them up if you need more examples.

File: src/direct-binary-upload-process.js (hunk 1 above)
Comment (mfrisbey) [reply src/direct-binary-upload-process.js:96]: Good catch, thank you. I've removed any signed URLs from this operation.

File: src/direct-binary-upload-process.js (hunk 2 above)
Comment (mfrisbey) [reply src/direct-binary-upload-process.js:130]: I double-checked that the url is a non-signed URL (so we should be ok here). I went ahead and changed the level to `debug`, though, just to be safe.

File: test/upload-result.test.js (hunk 3 above)
Comment (mfrisbey) [reply test/upload-result.test.js:24]: Good suggestion, thank you. I'll look into the documentation.

**Pull request closed without merging**